### PR TITLE
Newmoule nextdenovo

### PR DESCRIPTION
**Pull request by scorreard**
## PR checklist

Closes #[2564](https://github.com/nf-core/modules/issues/2564) 

- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [ x Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

**Comment by scorreard** [comment]
The tests fails because of a lack of compute capacity in Github.
```
[772 WARNING] 2023-05-03 00:24:43 Adjust -p from 2 to 0, -w from 5000000 to 5000000, logical CPUs:2, available RAM:~5G, use -a to disable automatic adjustment.
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/nextdenovo/lib/ctg_cns.py", line 263, in <module>
    main(args)
  File "/usr/local/lib/python3.9/site-packages/nextdenovo/lib/ctg_cns.py", line 192, in main
    pool = Pool(args.process, initializer=start)
  File "/usr/local/lib/python3.9/multiprocessing/context.py", line 119, in Pool
    return Pool(processes, initializer, initargs, maxtasksperchild,
  File "/usr/local/lib/python3.9/multiprocessing/pool.py", line 205, in __init__
    raise ValueError("Number of processes must be at least 1")
ValueError: Number of processes must be at least 1
```

The automatic adjustment is not happening when I run the script on GitLab.
The -a option that they mention is not available (I tried using it and it gave me a "unknown parameter" error).

I don't see a way around this issue

**Comment by SPPearce** [comment]
Hi @scorreard , what is the status of this PR? Do you need some help to try and get it passing?

**Comment by scorreard** [comment]
Hi @SPPearce, yes I could use a review.
This would be useful for the GenomeAssembler pipeline, but it has been quiet lately.

I also know that it is a lot of work to approve a module that fails the test. But I'm confident that the code is good.

**Comment by SPPearce** [comment]
Ok, we can try and figure out a way forward.
If we can't test it properly, one option would be to skip the tests on the CI by using a stub, which is done for example in modules that require a large test database or otherwise overwhelm the CI limits.

**Comment by SPPearce** [comment]
The test works fine on gitpod, so it is something about GitHub test environment.
Can you downsize the input further? Change the parameters to the tool so it doesn't have to compute as much?

**Comment by scorreard** [comment]
I can try to do that during the hackathon, I won't have time before. :)

**Comment by SPPearce** [comment]
That's fine, I'm just trying to help get some of the PRs finished off

File: modules/nf-core/nextdenovo/meta.yml [hunk 1]
```diff
@@ -0,0 +1,70 @@
+---
+# yaml-language-server: $schema=https://raw.githubusercontent.com/nf-core/modules/master/modules/yaml-schema.json
+name: "nextdenovo"
+description: string graph-based de novo assembler for long reads (CLR, HiFi and ONT)
+keywords:
+  - assembler
+  - long reads
+  - correaction
```
Comment (SPPearce) [inline modules/nf-core/nextdenovo/meta.yml:8]: Is this a word?

File: modules/nf-core/nextdenovo/meta.yml [hunk 2]
```diff
@@ -0,0 +1,70 @@
+---
+# yaml-language-server: $schema=https://raw.githubusercontent.com/nf-core/modules/master/modules/yaml-schema.json
+name: "nextdenovo"
+description: string graph-based de novo assembler for long reads (CLR, HiFi and ONT)
+keywords:
+  - assembler
+  - long reads
```
Comment (SPPearce) [inline modules/nf-core/nextdenovo/meta.yml:7]: Hmm, does the linter allow two words?

File: modules/nf-core/nextdenovo/meta.yml [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+      description: task need to run, correct = only do the correction step, assemble = only do the assembly step, all = correct + assemble.
+      pattern: "{all, correct, assemble}"
+  - input_type:
+      type: string
+      description: input reads type
+      pattern: "{raw, corrected}"
+  - genome_size:
+      type: string
+      description: estimated genome size, suffix K/M/G recognized
+      pattern: "{3g}"
```
Comment (SPPearce) [inline modules/nf-core/nextdenovo/meta.yml:44]: What is this pattern actually matching?

File: modules/nf-core/nextdenovo/main.nf [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    nextDenovo run.cfg
+
+    #Data organization
+    gzip -c ${prefix}/02.cns_align/01.seed_cns.sh.work/seed_cns*/cns.fasta > ${prefix}_cns.fasta.gz
+    gzip -c ${prefix}/03.ctg_graph/nd.asm.fasta > ${prefix}_nd.asm.fasta.gz
+    mv ${prefix}/03.ctg_graph/nd.asm.fasta.stat ${prefix}_nd.asm.fasta.stat
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        nextDenovo : \$(echo \$(nextDenovo --version 2>&1) | sed 's/^.*nextDenovo //; s/Using.*\$//' )
```
Comment (SPPearce) [inline modules/nf-core/nextdenovo/main.nf:73]: What exactly is this sed command doing, with two parts to it? Could they be combined into one?

File: modules/nf-core/nextdenovo/main.nf [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        deltmp = yes
+        parallel_jobs = ${task.cpus}
+        input_type = $input_type
+        read_type = ${read_type}
+        input_fofn = input.fofn
+        workdir = $prefix
+        $args_general
+
+        [correct_option]
+        read_cutoff = 1k
```
Comment (SPPearce) [inline modules/nf-core/nextdenovo/main.nf:52]: Is this hard coding something?

File: modules/nf-core/nextdenovo/main.nf [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        genome_size = $genome_size
+        sort_options = -t ${task.cpus} ${args_sort_options}
+        minimap2_options_raw = -t ${task.cpus} ${args_minimap2_options_raw}
+        pa_correction = 5
```
Comment (SPPearce) [inline modules/nf-core/nextdenovo/main.nf:56]: And this?

File: modules/nf-core/nextdenovo/main.nf [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    def args_sort_options = task.ext.args_sort_options ?: ''
+    def args_minimap2_options_raw = task.ext.args_minimap2_options_raw ?: ''
+    def args_minimap2_options_cns = task.ext.args_minimap2_options_cns ?: ''
+    def args_nextgraph_options = task.ext.args_nextgraph_options ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    """
+    #Prepare the input file
+    ls ${reads} > input.fofn
+
+    #Prepare the conf file run.cfg
```
Comment (SPPearce) [inline modules/nf-core/nextdenovo/main.nf:37]: Ewww, tools that review config files are tricky 😢

File: modules/nf-core/nextdenovo/main.nf [hunk 8]
```diff
@@ -0,0 +1,94 @@
+process NEXTDENOVO {
+    tag "$meta.id"
+    label 'process_medium'
+
+    conda "bioconda::nextdenovo=2.5.2"
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/nextdenovo%3A2.5.2--py39h20169af_0':
```
Comment (SPPearce) [inline modules/nf-core/nextdenovo/main.nf:7]:
```suggestion
        'https://depot.galaxyproject.org/singularity/nextdenovo:2.5.2--py39h20169af_0':
```

File: modules/nf-core/nextdenovo/main.nf [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    tuple val(meta), path("*cns.fasta.gz")  , emit: corrected_reads
+    tuple val(meta), path("*asm.fasta.gz")  , emit: assembly
+    tuple val(meta), path("*asm.fasta.stat"), emit: stat
+    path "versions.yml"                     , emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args_general = task.ext.args_general ?: ''
```
Comment (SPPearce) [inline modules/nf-core/nextdenovo/main.nf:27]: Hmmm. The general consensus is that we should only use `args`, `args2`, `args3` etc. Personally I feel this is much more readable like this, but it makes the whole set of modules harder to maintain when there are lots of different named `task.ext.X` values.

File: modules/nf-core/nextdenovo/main.nf [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+    stub:
+    def args_general = task.ext.args_general ?: ''
+    def args_sort_options = task.ext.args_sort_options ?: ''
+    def args_minimap2_options_raw = task.ext.args_minimap2_options_raw ?: ''
+    def args_minimap2_options_cns = task.ext.args_minimap2_options_cns ?: ''
+    def args_nextgraph_options = task.ext.args_nextgraph_options ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    """
+    touch ${prefix}.cns.fasta.gz
```
Comment (SPPearce) [inline modules/nf-core/nextdenovo/main.nf:85]: This stub is making differently named files.

File: modules/nf-core/nextdenovo/meta.yml (hunk 2 above)
Comment (scorreard) [reply modules/nf-core/nextdenovo/meta.yml]: It appear it does. Still, I added a dash to be sure

File: modules/nf-core/nextdenovo/meta.yml (hunk 3 above)
Comment (scorreard) [reply modules/nf-core/nextdenovo/meta.yml:44]: It is an example of what it should look like for a genome of 3Gb. Should it be done differently?

**Comment by SPPearce** [comment]
I think the pytest issues should be resolved by this: https://github.com/nf-core/modules/pull/4118
If you rename `modules/nf-core/custom/getchromsizes/tests/test.yml` to `modules/nf-core/custom/getchromsizes/tests/tags.yml` it should hopefully run for you too.

**Comment by scorreard** [comment]
Closing this PR without merging as a lot of changes were made to module creation and nf-test, I think it would be better to strat from scratch.

**Pull request closed without merging**